uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+use shared_algorithm_lens_accessors :: * ;
+
+# [derive (Clone , Debug , Default , candid :: CandidType , serde :: Deserialize , serde :: Serialize)]
+pub struct LensValue {
+  pub address: String,
+  pub current_price: f32,
+  pub range_top: f32,
+  pub range_bottom: f32,
+  pub edpr: f32
+}
+
+pub async fn calculate (targets : Vec < String >) -> LensValue {
```

Comment:
algorithm_lens is arbitrary rust logic, so you can write test code. Also, since your logic is not short, but rather a combination of several calculations, you should show that these are sound in your test code.

In showcase, tests are written for both short and long code.
https://github.com/horizonx-tech/chainsight-showcase/blob/main/price_feed/price_wbtc/src/logics/lens_wbtcusd/src/lib.rs
https://github.com/horizonx-tech/chainsight-showcase/blob/main/volatility_index_spx/src/logics/lens_vix_spx/src/calc/variance.rs
https://github.com/horizonx-tech/chainsight-showcase/blob/main/volatility_index_spx/src/logics/lens_vix_spx/src/calc/vix.rs

I would like to add that it would be nice if we could test and guarantee some formulas, not the entire code from the beginning to the end.
For example
- Calculate amount,tvl from tick
- Calculation of edr from tvl and fee, etc.
The pattern of the order of token0 and token1 should also be checked.

Reply:
Thanks for pointing it out. It's been informally checked in the Rust Playground (various links in this thread) and by deployment, with successful results. Testing as per the examples you've given is beyond anything we've discussed so far, so I'm going to give a "not willing" in case you're needing this specified as an answer.

Not willing.

Reply:
What we are guaranteeing here follows what we have been discussing. However, this is a general requirement in the completed code.
If it were simple or obvious, it could be skipped, but the amount of discussion so far makes it clear that this is necessary. In particular, the difference between token0,token1 and the EDR calculation would not raise similar questions if they were secured by testing.

Reply:
All the answers are in the preceding discussion.

Reply:
![image](https://github.com/horizonx-tech/chainsight-showcase/assets/37236322/e8810f5b-4405-4a39-81eb-122449160a5f)

From https://info.uniswap.org/#/pools/0xc2e9f25be6257c210d7adf0d4cd6e3e881ba25f8:

  Range is approximately 16 tick brackets.
  Value locked within the range should approximate to 16 * 73 ETH = 1168 ETH.

  current_sqrt_price = (1/2529.07)^0.5 = 0.0198847
  range_top.sqrt() = (1/2469.1)^0.5 = 0.0201247
  range_bottom.sqrt() = (1/2701.61)^0.5 = 0.0192392

From https://etherscan.io/address/0xc2e9f25be6257c210d7adf0d4cd6e3e881ba25f8#readContract#F5:

  current_tick_liquidity = 1230007656009839982219288

Using calculator app to work out expected values from the code:

  let amount0 = current_tick_liquidity as f32
    * (range_top.sqrt() - current_sqrt_price as f32)
    / (current_sqrt_price as f32 * range_top.sqrt())
    / f32::powf(10.0, t0_decimals as f32);
  = 1230007656009839982219288*(0.0201247-0.0198847)
    / 0.0198847*0.0201247 / 1e18
  = 295.201837/0.0004001736
  = 737,684.437

  let amount1 = current_tick_liquidity as f32
    * (current_sqrt_price as f32 - range_bottom.sqrt())
    / f32::powf(10.0, t1_decimals as f32);
  = 1230007656009839982219288*(0.0198847-0.0192392) / 1e18
  = 793.9699

  let mut est_tvl_in_range = amount0 * current_price + amount1;
  = 737,684.437 * (1/2529.07)    // noting that current_price represents ETH per DAI
    + 793.9699
  = 737684.437/2529.07+793.9699
  = 1,085.65 // expressed in ETH

This is sufficiently close to the earlier value of 1168 ETH to be believable.

Reply:
Shown above is not a formal test, but just an informal calculation to give a rough indication as to whether the [current code](https://github.com/timk11/chainsight-showcase/blob/2a37216167d7eb76ab16e230b94729bd094e4b10/uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs#L56-L63) is on track.

Presented here just for information, and to help with my own thinking around developing some formal tests.

Reply:
Thanks for your consideration.
I have decided not to make this mandatory. I have decided to consider your time.
(If you can, please continue to write the test code)